Passkeys: Supporting WebAuthn Conditional UI (#24305)
keycloak/keycloak: docs/documentation/server_admin/topics/authentication/passkeys.adoc

Both synced passkeys and device-bound passkeys can be used for both Same-Device and Cross-Device Authentication (CDA).
However, passkeys operations success depends on the user's environment. Make sure which operations can succeed in https://passkeys.dev/device-support/[the environment].
====

[[_passkeys-conditional-ui]]
==== Passkey Authentication with Conditional UI

Passkey Authentication with Conditional UI can authenticate a user with its passkey in the same way as in xref:_webauthn_loginless[LoginLess WebAuthn].
This authentication shows a user a list of passkeys stored on a device where the user runs a browser.
Therefore, the user can select one of the passkeys in the list to authenticate them. Compared with xref:_webauthn_loginless[LoginLess WebAuthn], the authentication improves the user's experience of authentication.

[NOTE]
====

> This authentication uses https://github.com/w3c/webauthn/wiki/Explainer:-WebAuthn-Conditional-UI/[WebAuthn Conditional UI].

Reviewer: ```suggestion
This authentication uses the https://github.com/w3c/webauthn/wiki/Explainer:-WebAuthn-Conditional-UI/[WebAuthn Conditional UI].
```

Author: Thank you. I will fix it as you suggested.

Therefore, this authentication success depends on the user's environment.
If the environment does not support WebAuthn Conditional UI, this authentication falls back to xref:_webauthn_loginless[LoginLess WebAuthn].
====

:tech_feature_name: Passkey Authentication
:tech_feature_setting: -Dkeycloak.profile.feature.passkeys=enabled
:tech_feature_id: passkeys
include::../templates/techpreview.adoc[]

.Procedure
===== Setup

Set up Passkey Authentication with Conditional UI as follows:

. (if not already present) Register a new required action for WebAuthn passwordless support. Use the steps described in <<_webauthn-register, Enable WebAuthn Authenticator Registration>>. Register the `Webauthn Register Passwordless` action.

> . Configure the `WebAuthn Passwordless Policy`. Perform the configuration in the Admin Console, `Authentication` section,  in the tab `Policies` ->  `WebAuthn Passwordless Policy`. You have to set *User Verification Requirement* to *required* and *Require Resident Key* to *Yes* when you configure the policy for loginless scenario. Note that since there isn't a dedicated Loginless policy it won't be possible to mix authentication scenarios with user verification=no/resident key=no and loginless scenarios (user verification=yes/resident key=yes). Storage capacity is usually very limited on passkeys meaning that you won't be able to store many resident keys on your pass key.

Reviewer: ```suggestion
. Configure the `WebAuthn Passwordless Policy`. Perform the configuration in the Admin Console, `Authentication` section,  in the tab `Policies` ->  `WebAuthn Passwordless Policy`. Set *User Verification Requirement* to *required* and *Require Resident Key* to *Yes* when you configure the policy for loginless scenario. Note that since there is no dedicated Loginless policy, it is impossible to mix authentication scenarios with user verification=no/resident key=no and loginless scenarios (user verification=yes/resident key=yes). Storage capacity is usually very limited on passkeys meaning that you cannot store many resident keys on your pass key.
```

Author: Thank you. I will fix it as you suggested.

. Configure the authentication flow. Create a new authentication flow, add the "WebAuthn Passwordless" execution and set the Requirement setting of the execution to *Required*

The final configuration of the flow looks similar to this:

.Passkey Authentication with Conditional UI flow
image:images/passkey-conditional-ui-flow.png[Passkey Authentication with Conditional UI flow flow]

> You can now add the required action `WebAuthn Register Passwordless` to a user, already known to {project_name}, to test this. The user with the required action configured will have to authenticate (with a username/password for example) and will then be prompted to register a passkey to be used for loginless authentication.

Reviewer: ```suggestion
Add the required action `WebAuthn Register Passwordless` to a user who is already known to {project_name} to test this. The user with the required action configured will have to authenticate (with a username/password for example) and will then be prompted to register a passkey to be used for loginless authentication.
```

Author: Thank you. I will fix it as you suggested.

====

:tech_feature_name: Passkey Authentication
:tech_feature_setting: -Dkeycloak.profile.feature.passkeys=enabled
:tech_feature_id: passkeys
include::../templates/techpreview.adoc[]

.Procedure
===== Setup

Set up Passkey Authentication with Conditional UI as follows:

. (if not already present) Register a new required action for WebAuthn passwordless support. Use the steps described in <<_webauthn-register, Enable WebAuthn Authenticator Registration>>. Register the `Webauthn Register Passwordless` action.

. Configure the `WebAuthn Passwordless Policy`. Perform the configuration in the Admin Console, `Authentication` section,  in the tab `Policies` ->  `WebAuthn Passwordless Policy`. Set *User Verification Requirement* to *required* and *Require discoverable credential* to *Yes* when you configure the policy for loginless scenario. Note that since there is no dedicated Loginless policy, it is impossible to mix authentication scenarios with user verification=no/discoverable credential=no and loginless scenarios (user verification=yes/discoverable credential=yes).
+

> NOTE: Storage capacity is usually very limited on hardware passkeys meaning that you cannot store many discoverable credentials on your passkey. However this limitation may be mitigated for instance if you use Android phone backed by Google account as a passkey device or iPhone backed by Bitwarden.

Reviewer: ```suggestion
NOTE: Storage capacity is usually very limited on hardware passkeys meaning that you cannot store many discoverable credentials on your passkey. However, this limitation may be mitigated for instance if you use an Android phone backed by a Google account as a passkey device or an iPhone backed by Bitwarden.
```

Author: Thank you. I fixed it as you suggested.

> . Configure the authentication flow. Create a new authentication flow, add the *Passkeys Conditional UI Authenticator* execution and set the Requirement setting of the execution to *Required*

Reviewer: ```suggestion
. Configure the authentication flow. Create a new authentication flow, add the *Passkeys Conditional UI Authenticator* execution and set the Requirement setting of the execution to *Required*.
```

Author: Thank you. I fixed it as you suggested.

+
The final configuration of the flow looks similar to this:
image:images/passkey-conditional-ui-flow.png[Passkey Authentication with Conditional UI flow flow]

> . Bind the flow above as *browser* authentication flow in the realm as described in the <<_webauthn-register, WebAuthn section above>>.

Reviewer: ```suggestion
. Bind the flow above as a *browser* authentication flow in the realm as described in the <<_webauthn-register, WebAuthn section above>>.
```

Author: Thank you. I fixed it as you suggested.

> The authentication flow above requires that user must already have passkey credential on his account to be able to login. This means that all users in the realm must have passkeys already set.

Author: Thank you. I fixed it as you suggested.

> That can be achieved for instance by enable user registration as described below.

Reviewer: ```suggestion
That can be achieved for instance by enabling user registration as described below.
```

Author: Thank you. I fixed it as you suggested.

===== Setup of the registration for passkeys conditional UI

. Enable <<con-user-registration_{context}, registration>> for your realm

. In the <<configuring-authentication_{context},authentication flows>> of the realm, select flow *registration* and switch the authenticator *Password validation* to *Disabled*.

> This means that newly registered users will not be required to create the passwords as for this example setup, we want the users to always use passkeys instead of the passwords.

Reviewer: ```suggestion
This means that newly registered users will not be required to create the passwords in this example setup. Users must always use passkeys instead of passwords.
```

Author: Thank you. I fixed it as you suggested.

> . Go back to *Required actions* sub-tab of the tab *Authentication* tab and find the `Webauthn Register Passwordless` action and mark it with *Set as default action*.

Reviewer: ```suggestion
.  Return to the *Required actions* sub-tab of the tab *Authentication* tab and find the `Webauthn Register Passwordless` action and mark it with *Set as default action*.
```

Author: Thank you. I fixed it as you suggested.

This means that it would be added to all new users after their registration.

> The alternative to the registration flow setup is to add the required action `WebAuthn Register Passwordless` to a user who is already known to {project_name}, to test this. The user with the required action configured will have to authenticate (with a username/password for example) and will then be prompted to register a passkey to be used for loginless authentication.

Reviewer: Not sure what "to test this" means.

Author: "to test this" is not needed so that I removed the words. Thank you.

[NOTE]
====

> We plan to improve the usability and allow to integrate conditional passkeys with the existing authenticators and forms like the default username / password form.

Reviewer: ```suggestion
We plan to improve the usability and allow integration of conditional passkeys with the existing authenticators and forms such as the default username / password form.
```

Author: Not sure what "to test this" means.

Author: Thank you. I fixed it as you suggested.
